test_bin/src/test_bin_ipc.rs

```diff
@@ -56,140 +81,108 @@ fn exec() -> NetResult<()> {
-            msg_id: "unique-id".to_string(),
-            data: json!("test data"),
-        })
-        .into(),
-    )?;
+    let mut node1 = spawn_connection(&n3h_path)?;
+    let mut node2 = spawn_connection(&n3h_path)?;
 
-    let handle_data;
+    println!("node1 path: {}", node1.dir);
+    println!("node2 path: {}", node2.dir);
 
-    // loop waiting for the request to handle a message
-    loop {
-        let z = receiver.recv()?;
+    let node1_state = node1.wait(Box::new(is_state))?;
+    let node2_state = node2.wait(Box::new(is_state))?;
 
-        if let Ok(wrap) = ProtocolWrapper::try_from(&z) {
-            match wrap {
-                ProtocolWrapper::HandleSend(m) => {
-                    handle_data = m;
-                    break;
-                }
-                _ => (),
-            }
+    let node1_id = {
+        if let ProtocolWrapper::State(s) = node1_state {
+            s.id
+        } else {
+            unimplemented!()
         }
+    };
 
-        println!("got: {:?}", z);
-    }
+    let node2_binding = {
+        if let ProtocolWrapper::State(s) = node2_state {
+            s.bindings[0].clone()
+        } else {
+            unimplemented!()
+        }
+    };
+
+    println!("connect node1 ({}) to node2 ({})", node1_id, node2_binding);
 
-    println!("got handleSend: {:?}", handle_data);
-
-    // hey, we got a message from ourselves!
-    // let's send ourselves a response
-    con.send(
-        ProtocolWrapper::HandleSendResult(MessageData {
-            dna_hash: handle_data.dna_hash,
-            to_agent_id: handle_data.from_agent_id,
-            from_agent_id: AGENT_ID.to_string(),
-            msg_id: handle_data.msg_id,
-            data: json!(format!("echo: {}", handle_data.data)),
+    node1.con.send(
+        ProtocolWrapper::Connect(ConnectData {
+            address: node2_binding,
         })
         .into(),
     )?;
 
-    // wait for the response to our original message
-    loop {
-        let z = receiver.recv()?;
-        println!("got: {:?}", z);
-
-        if let Ok(wrap) = ProtocolWrapper::try_from(&z) {
-            match wrap {
-                ProtocolWrapper::SendResult(m) => {
-                    println!("Got Result! : {:?}", m);
-
-                    assert_eq!(
-                        "echo: \"test data\"".to_string(),
-                        m.data.as_str().unwrap().to_string(),
-                    );
-
-                    break;
-                }
-                _ => (),
-            }
-        }
-
-        println!("got: {:?}", z);
-    }
+    let connect_result_1 = node1.wait(Box::new(is_peer_connected))?;
+    println!("got connect result 1: {:?}", connect_result_1);
 
-    // yay, everything worked
-    println!("test complete");
+    let connect_result_2 = node2.wait(Box::new(is_peer_connected))?;
+    println!("got connect result 2: {:?}", connect_result_2);
 
-    // shut down the P2pNetwork instance
-    con.stop()?;
+    node1.drop();
+    node2.drop();
```

Comment:
Why are you dropping those by hand?

Reply:
erm... sorry, bad name. changing to `stop` for now. `stop` on P2pNetwork right now takes `self` (moved, not a reference). I'd like to fix that, but as a separate effort not part of this.